tsdb/importer/importer.go

```diff
@@ -0,0 +1,602 @@
+	scanner.Split(sampleStreamer(buf))
+
+	for scanner.Scan() {
+		if currentPassCount == 0 {
+			blocks = make([][]*tsdb.MetricSample, len(metas))
+		}
+
+		encSample := scanner.Bytes()
+		decBuf := bytes.NewBuffer(encSample)
+		sample := tsdb.MetricSample{}
+		if err := gob.NewDecoder(decBuf).Decode(&sample); err != nil {
+			level.Error(logger).Log("msg", "failed to decode current entry returned by file scanner", "err", err)
+			return err
+		}
+
+		// Find what block this sample belongs to.
+		var blockIndex int
+		if len(metas) == 1 || sample.Timestamp < dbMint {
+			blockIndex = 0
+		} else if sample.Timestamp >= dbMaxt {
+			blockIndex = len(metas) - 1
+		} else {
+			blockIndex = getBlockIndex(sample.Timestamp, metas)
+			if blockIndex == -1 {
+				return NoBlockMetaFoundError
+			}
+		}
+
+		meta := metas[blockIndex]
+		meta.mint = value.MinInt64(meta.mint, sample.Timestamp)
+		meta.maxt = value.MaxInt64(meta.maxt, sample.Timestamp)
+		meta.count++
+
+		blocks[blockIndex] = append(blocks[blockIndex], &sample)
+
+		currentPassCount++
+		// Have enough samples to write to disk.
+		if currentPassCount == maxSamplesInMemory {
+			if err := flushBlocks(dir, blocks, metas, maxBlockChildren, logger); err != nil {
+				return err
+			}
+			// Reset current pass count.
+			currentPassCount = 0
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+
+	// Flush any remaining samples.
+	if err := flushBlocks(dir, blocks, metas, maxBlockChildren, logger); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// flushBlocks writes the given blocks of samples to disk, and updates block metadata information.
+func flushBlocks(
+	dir string,
+	toFlush [][]*tsdb.MetricSample,
+	metas []*newBlockMeta,
+	maxBlockChildren int,
+	logger log.Logger,
+) error {
+	for blockIdx, block := range toFlush {
+		// If current block is empty, nothing to write.
+		if len(block) == 0 {
+			continue
+		}
+		meta := metas[blockIdx]
+
+		// Put each sample into the appropriate block.
+		bins, binTimes := binSamples(block, meta, tsdb.DefaultBlockDuration)
+		for binIdx, bin := range bins {
+			if len(bin) == 0 {
+				continue
+			}
+			start, end := binTimes[binIdx].start, binTimes[binIdx].end
+			path, err := tsdb.CreateBlock(bin, dir, start, end, logger)
+			if err != nil {
+				return err
+			}
+			child := &newBlockMeta{
+				index:     meta.index,
+				count:     0,
+				mint:      start,
+				maxt:      end,
+				isAligned: meta.isAligned,
+				dir:       path,
+			}
+			meta.children = append(meta.children, child)
+		}
+		if maxBlockChildren > 0 && len(meta.children) >= maxBlockChildren {
+			if err := compactMeta(dir, meta, logger); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// sampleStreamer returns a function that can be used to parse an OpenMetrics compliant
+// byte array, and return each token in the user-provided byte buffer.
+func sampleStreamer(buf *bytes.Buffer) func([]byte, bool) (int, []byte, error) {
```

Comment:
I am not fan of this function.

1. Why we encode twice? We parse, then encode in buffer with `enc := gob.NewEncoder(buf)`. Why we do that?
2. textparse.Parse is not well suited as time series interface ): It looks not well suited for formats that allows to give you samples in bulk. I think we need to have something different anyway ):